store/store_messaging.go

```diff
@@ -0,0 +1,130 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package store has support to use the Ubuntu Store for querying and downloading of snaps, and the related services.
+package store
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+)
+
+const (
+	messagesEndpointPath = "v2/messages"
+)
+
+// PollMessagesRequest contains parameters for polling & sending messages to the Store.
+type PollMessagesRequest struct {
+	// Token of last successfully stored request-message.
+	// Acknowledges all messages up to and including the specified token.
+	After string `json:"after,omitempty"`
+
+	// Controls the operation mode:
+	//   - limit > 0: Poll for up to limit messages (server may return fewer)
+	//   - limit = 0: Don't return messages (useful for ack-only, response-only, or status checks)
+	Limit int `json:"limit"`
+
+	// The response messages to send to the Store.
+	Messages []Message `json:"messages,omitempty"` // Response messages to send
+}
+
+// PollMessagesResponse contains request-message messages received from the store with their tokens.
+// PollMessagesResponse contains messages and queue status from the Store.
+//
+// When:
+//   - limit > 0: messages contains up to limit request messages
+//   - limit = 0: messages is omitted, only queue status is returned
+type PollMessagesResponse struct {
+	// Request messages with their acknowledgement tokens.
+	Messages []MessageWithToken `json:"messages,omitempty"`
+
+	// Total unacknowledged messages in the device's queue.
+	TotalPendingMessages int `json:"total-pending-messages"`
+}
+
+// messageQueueError contains error responses from the message queue.
+type messageQueueError struct {
+	ErrorList []struct {
+		Code    string `json:"code"`
+		Message string `json:"message"`
+	} `json:"error-list"`
+}
+
+// Message represents a message and its format.
+type Message struct {
+	Format string `json:"format"` // e.g. assertion
+	Data   string `json:"data"`   // Encoded assertion
+}
+
+// MessageWithToken is a request-message with its acknowledgement token.
+// The token should be echoed back in a subsequent poll's After field once the
+// message has been successfully received and stored.
+type MessageWithToken struct {
+	Message
+	Token string `json:"token"`
+}
+
+// PollMessages polls the store's /v2/messages endpoint, sending response-message messages
+// and acknowledging received request-message messages.
+func (s *Store) PollMessages(ctx context.Context, req *PollMessagesRequest) (*PollMessagesResponse, error) {
+	if req == nil {
+		return nil, fmt.Errorf("poll request cannot be nil")
```

Comment:
Same idea, IMO it's clearer if we call this a `messages request` or a `fetch messages request`